### PR TITLE
Add %noun-autodiff mark with automatic diffing algorithm

### PR DESCRIPTION
**Pull request by ilyakooo0**
This implementation is heaily inspired by the paper ["An Efficient Algorithm for Type-Safe Structural Diffing"](https://github.com/urbit/urbit/files/11817181/An_efficient_algorithm_for_type-safe_structural_di.pdf).

The test requires #6558 to be merged.

The current implementation only diffs the tree structure (cells). (Atoms are not diffed, and any new atoms are present in the patches as-is.)

## Potential future work

1. Add atom diffing.
2. Optimize the implementation
3. Add patch merging
4. Rename things to cute urbit names

### CODE REVIEW

File: pkg/base-dev/mar/noun.hoon [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ::
 ::::  /hoon/noun/mar
   ::
-/?    310
```
Comment (ilyakooo0) [inline pkg/base-dev/mar/noun.hoon:4]: I should probably put this back

File: pkg/base-dev/mar/noun.hoon [hunk 2]
```diff
@@ -14,9 +14,9 @@
 ++  grad
   |%
   ++  form  %noun
-  ++  diff  |=(* +<)
-  ++  pact  |=(* +<)
-  ++  join  |=([* *] *(unit *))
-  ++  mash  |=([[ship desk *] [ship desk *]] `*`~|(%noun-mash !!))
+  ++  diff  diff:noun-diff
+  ++  pact  apply:noun-diff
```
Comment (ilyakooo0) [inline pkg/base-dev/mar/noun.hoon:18]: This also doesn't compile at the moment

**Comment by ilyakooo0** [comment]
BTW I can make the algorithm ~1.5x faster at the cost of larger diffs

File: tests/lib/noun-diff.hoon [hunk 3]
```diff
@@ -0,0 +1,22 @@
+/+  quiz, *test, noun-diff
+
```
Comment (belisarius222) [inline tests/lib/noun-diff.hoon:2]: blank lines

File: pkg/base-dev/lib/noun-diff.hoon [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    [%diff [%hole ~] [%hole ~]]
+  +$  patch
+    $%
+      [%cell lhs=patch rhs=patch]
+      [%diff ^diff]
+    ==
+--
+::
+|%
+  +$  del-diff
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:35]:
```hoon
|%
+$  del-diff
  $%  [%hole @]
       ::  ...
  ==
++  ins-diff
  ::  ...
::  ...
```

File: pkg/base-dev/lib/noun-diff.hoon [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+          [%cell del:pulled-lhs del:pulled-rhs]
+          [%cell ins:pulled-lhs ins:pulled-rhs]
+    ==
+  ++  is-subtree  
+    |=  [tree=* subtree=*]
+    ~+
+    ^-  ?
+    ?:  =(tree subtree)  &
+    ?@  tree  |
+    ?|
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:164]:
```hoon
    ?|  (is-subtree ...)
        (is-subtree ...)
    ==
```

File: pkg/base-dev/lib/noun-diff.hoon [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+      (is-subtree -.tree subtree)
+      (is-subtree +.tree subtree)
+    ==
+  ++  oracle
+    |=  [a=* b=*]
+    |=  subtree=*
+    ^-  (unit @)
+    ?: 
+    ?&
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:173]: same as the `?|` earlier

File: pkg/base-dev/lib/noun-diff.hoon [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    ?@  tree  |
+    ?|
+      (is-subtree -.tree subtree)
+      (is-subtree +.tree subtree)
+    ==
+  ++  oracle
+    |=  [a=* b=*]
+    |=  subtree=*
+    ^-  (unit @)
+    ?:
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:172]:
```hoon
?:  ?&  ...
    ==
  `(mug subtree)`
~
```

or 
```hoon
?.  ?&  ...
    ==
  ~
`(mug subtree)`

```

File: pkg/base-dev/lib/noun-diff.hoon [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    ~+
+    ^-  ins-diff
+    =/  hash  (oracle subtree)
+    ?^  hash  [%hole +.hash subtree]
+    ?@  subtree  [%atom subtree]
+    [%cell (extract-ins oracle -.subtree) (extract-ins oracle +.subtree)]
+  ++  ins 
+    |=  [diff=final-ins-diff var-map=(map @ *)]
+    ^-  *
+    ?-  -.diff
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:198]: These are the two standard styles for `?-`s:

For long lines, use this:
```hoon
?-    -.diff
    %atom
  +.diff
::
    %cell
  [$(diff +<.diff) $(diff +>.diff)]
::
    ::  ...
==
```
For shorter lines, use this:
```hoon
?-  -.diff
  %atom  +.diff
  %cell  [$(diff +<.diff) $(diff +>.diff)]
  %hole  (~(got by var-map) +.diff)
==
```

File: pkg/base-dev/lib/noun-diff.hoon [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+      %cell
+        [$(diff +<.diff) $(diff +>.diff)]
+      %hole
+        (~(got by var-map) +.diff)
+    ==
+  ++  del  
+    |=  [diff=del-diff noun=*]  
+    ^-  (map @ *)
+    |^  (go diff noun ((map @ *) ~))
+      ++  go
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:210]:
```hoon
|^  (go diff noun ((map @ *) ~))
++  go
  |=  ...
---
```

**Comment by ilyakooo0** [comment]
@belisarius222 CI seems to pass. I guess I no longer need to extract the tests into a separate PR.

**Comment by joemfb** [comment]
@ilyakooo0 I was surprised to see that. Looked around a little more and ... #6705. I'll take a look tomorrow.

File: pkg/base-dev/lib/noun-diff.hoon [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+  =/  ins  (extract-ins (oracle old new) new)
+  =/  allowed-holes  (~(int in (find-del-holes del)) (find-ins-holes ins))
+  =.  del  (filter-del-holes allowed-holes del)
+  =/  ins  (filter-ins-holes allowed-holes ins)
+  =/  closed-patch  (closure (gcp [del ins]))
+  ?>  =(& +.closed-patch)
+  -.closed-patch
+++  apply  
+  |=  [patch=_id noun=*]
+  ?-  -.patch
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:16]:
```hoon
?-    -.patch
    %diff
    =/  var-map  (del del.patch noun)
    (ins ins.patch var-map)
::
    %cell
  ?>  ?=(^ noun)
  [$(patch lhs.patch, noun -.noun) $(patch rhs.patch, noun +.noun)]
==
```

File: pkg/base-dev/lib/noun-diff.hoon [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+  ==
+++  empty-set  (silt `(list @)`~)
+++  find-del-holes
+  |=  diff=del-diff
+  ~+
+  ^-  (set @)
+  ?-  -.diff
+    %hole  (silt ~[+.diff])
+    %cell
+      (~(uni in $(diff lhs.diff)) $(diff rhs.diff))
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:69]: move this to the previous line

File: pkg/base-dev/lib/noun-diff.hoon [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+  ==
+++  pull-diff
+  |=  =patch
+  ^-  [%diff del=del-diff ins=final-ins-diff]
+  ?-  -.patch
+    %diff  patch
+    %cell
+      =/  pulled-lhs  $(patch lhs:patch)
+      =/  pulled-rhs  $(patch rhs:patch)
+      :+
```
Comment (belisarius222) [inline pkg/base-dev/lib/noun-diff.hoon:148]:
```hoon
:+  %diff
  [%cell del:pulled-lhs del:pulled-rhs]
[%cell ins:pulled-lhs ins:pulled-rhs]
```

or 
```hoon
:*  %diff 
    [%cell del:pulled-lhs del:pulled-rhs]
    [%cell ins:pulled-lhs ins:pulled-rhs]
==
```

**Review by belisarius222 — APPROVED** [review]

**Pull request merged**